shared/util/quit-helper.desktop.js

```diff
@@ -3,18 +3,21 @@ import {ipcRenderer, ipcMain, app} from 'electron'
 import {quit} from '../../desktop/app/ctl'
 import {hideDockIcon} from '../../desktop/app/dock-icon'
 
-export type Context = {type: 'uiWindow'} | {type: 'mainThread'} | {type: 'quitButton'}
+export type Context = 'uiWindow' | 'mainThread' | 'quitButton' | 'beforeQuit'
+export type Action = 'closePopups' | 'quitMainWindow' | 'quitApp'
 
-export type Action = {type: 'closePopups'} | {type: 'quitMainWindow'} | {type: 'quitApp'}
+// Don't do beforeQuit path if we're really quitting
+let isQuitting = false
```

Comment:
before-quit is a special case. if the user chooses quit from the doc this is basically the first event we get, so we do want a full quit in this case but other flows where we quit (like through the widget) will also emit this event. we don't want to call quit, get before-quit, then do quit (or whatever else flow it does) so instead when we actually quit we stop this whole mechanism